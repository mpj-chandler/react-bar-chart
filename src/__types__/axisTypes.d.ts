import { SeriesData, NumericDataPoint } from './seriesTypes';
import { Padding } from './styling';
import { Placement } from '../enums/Placement';

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
    titlePos?: {
        xPos: number,
        yPos: number
    }
}

export interface AxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}

