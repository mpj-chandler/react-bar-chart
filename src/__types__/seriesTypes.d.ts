import DataType from '../enums/DataType';
import { Padding } from './styling';
import { AxisRange } from './axisTypes';
import { Placement } from '../enums/Placement';

export interface DataPoint {
    tag?: string;
    y0?: number | null;
    y1: number | null;
}

export interface NamedDataPoint extends DataPoint {
    x: string;    
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
}

export interface NumericDataPoint extends DataPoint {
    x: number;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}