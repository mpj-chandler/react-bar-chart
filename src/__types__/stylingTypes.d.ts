interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export default Padding;

