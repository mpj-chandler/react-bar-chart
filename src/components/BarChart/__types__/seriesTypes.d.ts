import DataType from "../enums/DataType";
import { Padding } from "./stylingTypes";
import { AxisRange } from "./axisTypes";
import Placement from "../enums/Placement";

export interface DataPoint {
    tag?: string;
}

export interface NamedDataPoint extends DataPoint {
    x: string;
    y: number;
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
    y: number;
}

export interface NumericDataPoint extends DataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}