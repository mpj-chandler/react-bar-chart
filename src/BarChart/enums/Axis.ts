enum Axis {
    XAxis,
    YAxis,
}

export default Axis;
