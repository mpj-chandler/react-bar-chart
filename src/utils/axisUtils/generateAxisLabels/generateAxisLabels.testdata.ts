import { SeriesData } from '../../../__types__/seriesTypes';
import DataType from '../../../enums/DataType';

export const dataOne: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 0,
                y1: 0.1,
            },
            {
                x: 1,
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 0,
                y1: 10.1,
            },
            {
                x: 1,
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 0,
                y1: 0.2,
            },
            {
                x: 1,
                y1: 6.6,
            },
        ],
    },
];

export const dataTwo: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: -1,
                y1: 0.1,
            },
            {
                x: 1,
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 0,
                y1: 10.1,
            },
            {
                x: 1,
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 0,
                y1: 0.2,
            },
            {
                x: 1,
                y1: 6.6,
            },
        ],
    },
];

const nullEntry: SeriesData = {
    seriesName: 'test-series-3',
    type: {
        x: DataType.NonNullNumeric,
        y: DataType.NonNullNumeric,
    },
    points: [
        {
            x: null,
            y1: 0.2,
        },
        {
            x: 1,
            y1: null,
        },
        {
            x: 2,
            y1: -1.2,
        },
        {
            x: 3,
            y1: 0.6,
        },
    ],
};

export const partiallyNullData: SeriesData[] = [ ...dataOne, nullEntry ];

export const malformedData: SeriesData[] = [
    {
        seriesName: 'test-series-3',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: null,
                y1: null,
            },
            {
                x: null,
                y1: null,
            },
            {
                x: null,
                y1: null,
            },
            {
                x: null,
                y1: null,
            },
        ],
    },
];

export const dataThree: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.Named,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 'A',
                y1: 0.1,
            },
            {
                x: 'B',
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.Named,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 'A',
                y1: 10.1,
            },
            {
                x: 'B',
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.Named,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 'A',
                y1: 0.2,
            },
            {
                x: 'B',
                y1: 6.6,
            },
        ],
    },
];

export const dataFour: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.Named,
            y: DataType.Named,
        },
        points: [
            {
                x: 'A',
                y1: 0.1,
            },
            {
                x: 'B',
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.Named,
            y: DataType.Named,
        },
        points: [
            {
                x: 'A',
                y1: 10.1,
            },
            {
                x: 'B',
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.Named,
            y: DataType.Named,
        },
        points: [
            {
                x: 'A',
                y1: 0.2,
            },
            {
                x: 'B',
                y1: 6.6,
            },
        ],
    },
];

export const badData: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.Named,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 'A',
                y1: 0.1,
            },
            {
                x: 'B',
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.NonNullNumeric,
            y: DataType.Named,
        },
        points: [
            {
                x: 'A',
                y1: 10.1,
            },
            {
                x: 'B',
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.Named,
            y: DataType.NonNullNumeric,
        },
        points: [
            {
                x: 'A',
                y1: 0.2,
            },
            {
                x: 'B',
                y1: 6.6,
            },
        ],
    },
];

export const unknownData: SeriesData[] = [
    {
        seriesName: 'test-series-1',
        type: {
            x: DataType.DateIndexed,
            y: DataType.DateIndexed,
        },
        points: [
            {
                x: 'A',
                y1: 0.1,
            },
            {
                x: 'B',
                y1: -3.5,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.DateIndexed,
            y: DataType.DateIndexed,
        },
        points: [
            {
                x: 'A',
                y1: 10.1,
            },
            {
                x: 'B',
                y1: 3.2,
            },
        ],
    },
    {
        seriesName: 'test-series-2',
        type: {
            x: DataType.DateIndexed,
            y: DataType.DateIndexed,
        },
        points: [
            {
                x: 'A',
                y1: 0.2,
            },
            {
                x: 'B',
                y1: 6.6,
            },
        ],
    },
];
