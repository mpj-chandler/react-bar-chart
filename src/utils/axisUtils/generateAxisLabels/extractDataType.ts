import Axis from '../../../enums/Axis';
import { SeriesData } from '../../../__types__/seriesTypes';
import DataType from '../../../enums/DataType';
export function extractDataType(seriesData: SeriesData[], axis: Axis): DataType {
    let type: DataType | null = null;
    seriesData.forEach((series: SeriesData) => {
        const newType = axis === Axis.XAxis ? series.type.x : series.type.y;
        if (type === null) {
            type = newType;
        } else if (type !== newType) {
            throw new Error('Error in axis calibration: all series data plotted on a given axis must be of the same type!');
        }
    });
    if (type === null) {
        throw new Error('Error in axis calibration: unable to determine what type of data is being plotted!');
    }

    return type;
}
