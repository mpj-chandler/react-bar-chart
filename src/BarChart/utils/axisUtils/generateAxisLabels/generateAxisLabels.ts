import Axis from '../../../enums/Axis';
import calculateTickInterval from '../calculateTickInterval/calculateTickInterval';
import { getXAxisRange } from '../getXAxisRange/getXAxisRange';
import { getYAxisRange } from '../getYAxisRange/getYAxisRange';
import { SeriesData, DataPoint, NamedDataPoint } from '../../../__types__/seriesTypes';
import { AxisRange } from '../../../__types__/axisTypes';
import DataType from '../../../enums/DataType';
import { extractDataType } from './extractDataType';
import { number } from 'prop-types';

function generateAxisLabels(seriesData: SeriesData[], axis: Axis):
    { dataType: DataType, labels: Array<{ index: number, label: string}> } {

    const dataType = extractDataType(seriesData, axis);

    switch (dataType) {
        case DataType.NonNullNumeric:
        case DataType.Numeric:
            return { dataType, labels: generateScaledNumericAxisLabels(seriesData, axis) };
        case DataType.Named:
            return { dataType, labels: generateNamedAxisLabels(seriesData, axis) };
        default:
            throw new Error('Error in axis calibration: unrecognised data type!');
    }
}

function generateScaledNumericAxisLabels(seriesData: SeriesData[], axis: Axis): Array<{ index: number, label: string}> {
    const labels: Array<{ index: number, label: string}> = [];
    let range: AxisRange;
    if (axis === Axis.XAxis) {
        range = getXAxisRange(seriesData);
    } else {
        range = getYAxisRange(seriesData);
    }

    const tickInterval = calculateTickInterval(range);
    const decimals = -1 * Math.min(0, Math.log10(tickInterval));
    let firstTickValue: number = 0;
    let index: number = 0;

    while (firstTickValue > range.min) {
        firstTickValue -= tickInterval;
    }
    let currentValue: number = firstTickValue;

    while (currentValue < range.max + tickInterval) {
        labels.push({ index, label: currentValue.toFixed(decimals) });
        currentValue += tickInterval;
        index += 1;
    }

    return labels;
}

function generateNamedAxisLabels(seriesData: SeriesData[], axis: Axis): Array<{ index: number, label: string}> {
    if (axis === Axis.YAxis) {
        throw new Error('Error in axis calibration: only numeric values currently allowed on y axis!');
    }

    const labels: Array<{ index: number, label: string}> = [];

    seriesData.forEach((data: SeriesData) => {
        data.points.forEach((point: DataPoint, index: number) => {
            labels.push({ index, label: (point as NamedDataPoint).x });
        });
    });

    return labels;
}

export default generateAxisLabels;
