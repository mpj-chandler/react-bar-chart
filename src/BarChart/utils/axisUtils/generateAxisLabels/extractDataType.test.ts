import { dataOne, dataTwo, dataThree, dataFour, badData } from './generateAxisLabels.testdata';
import { extractDataType } from './extractDataType';
import Axis from '../../../enums/Axis';
import { SeriesData } from '../../../__types__/seriesTypes';

describe('extractDataType', () => {

    [dataOne, dataTwo, dataThree, dataFour].forEach((data: SeriesData[], index) => {
        it(`it extracts the correct data type for data-${index + 1}`, () => {
            expect(extractDataType(dataOne, Axis.XAxis)).toMatchSnapshot();
            expect(extractDataType(dataOne, Axis.YAxis)).toMatchSnapshot();
        });
    });

    it('it throws an error when no data is available', () => {
        expect(() => extractDataType([], Axis.XAxis)).toThrowErrorMatchingSnapshot();
        expect(() => extractDataType([], Axis.YAxis)).toThrowErrorMatchingSnapshot();
    });

    it('it throws an error with inconsistent data', () => {
        expect(() => extractDataType(badData, Axis.XAxis)).toThrowErrorMatchingSnapshot();
        expect(() => extractDataType(badData, Axis.YAxis)).toThrowErrorMatchingSnapshot();
    });
});
