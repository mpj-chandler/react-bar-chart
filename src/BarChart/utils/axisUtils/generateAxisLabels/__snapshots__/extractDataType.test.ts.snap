// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractDataType it extracts the correct data type for data-1 1`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-1 2`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-2 1`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-2 2`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-3 1`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-3 2`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-4 1`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it extracts the correct data type for data-4 2`] = `"NON_NULL_NUMERIC"`;

exports[`extractDataType it throws an error when no data is available 1`] = `"Error in axis calibration: unable to determine what type of data is being plotted!"`;

exports[`extractDataType it throws an error when no data is available 2`] = `"Error in axis calibration: unable to determine what type of data is being plotted!"`;

exports[`extractDataType it throws an error with inconsistent data 1`] = `"Error in axis calibration: all series data plotted on a given axis must be of the same type!"`;

exports[`extractDataType it throws an error with inconsistent data 2`] = `"Error in axis calibration: all series data plotted on a given axis must be of the same type!"`;
